### ISSUE TITLE
Add analyzer to force ConfigureAwait(false) on all awaits

### ISSUE THREAD

**Issue by JamesNK** [dotnet/aspnetcore#31132]
We don't add `ConfigureAwait(false)` to ASP.NET Core code because there is no sync context. Microsoft.Extensions.* code could execute in client apps. Not using `ConfigureAwait(false)` can create deadlocks when a user mixes sync with async, e.g. https://github.com/aspnet/Extensions/issues/999.

Consider introducing an analyzer to enforce `ConfigureAwait(false)` everywhere. A quick search of this repos source code displays many non-test awaits that don't follow best practice.

I just had some good success using https://www.nuget.org/packages/ConfigureAwaitChecker.Analyzer/

**Comment by javiercn** [comment]
@JamesNK is there anything for us to do here?

Should we close this? If not, do you want to bring it for someone to pick it up in the DoI?

**Comment by JamesNK** [comment]
I don't know what Andrew ment when he said this is mostly done now. I don't think we've done anything 😬 

Basically I think we should turn on CA2007 for all our libraries that can run outside of ASP.NET Core. We're writing code without `ConfigureAwait(false)` because we're assuming that there is never a sync context, but some of our libraries can be used in WinForms or WPF. That's created some bugs.

We should come up with a test to apply the analyzer setting across the repo. All projects that are published as NuGet packages could be a place to start, and then figure out how to narrow down from their if nessessary.

Note that there are some projects where the rule will be turned off. For example, the SignalR client has a different strategy to avoid sync context issues (@BrennanConroy can you confirm, its been a couple of years since I looked at the .NET SignalR client code).

tldr; DoI should pick this up

**Comment by BrennanConroy** [comment]
> I don't know what Andrew ment when he said this is mostly done.

I think Andrew meant we had added `.ConfigureAwait(false)` to most of the Extensions code.

> For example, the .NET SignalR client has a different strategy to avoid sync context issues

Yeah, we have the weird `ForceAsync` method. It's very error prone though as you need to remember to add it to any new public top-level methods. We might want to discuss possibly changing that.

**Comment by pranavkm** [comment]
I'm going to consider this resolved based on https://github.com/dotnet/aspnetcore/pull/39946.
